Add unit tests for authn-k8s

doodlesbykumbi opened the pull request:
Add unit for configuring k8s api url with sub path. This unit test introduces a pattern that uses a test server that mimics the kubernetes api, and makes use of request specs.

Running the tests outputs

```
AuthenticateController
  #authenticate
    k8s api access contains subpath
[authn-k8s test server] Server running with port=1234, @subpath=/some/path
2022-06-02 14:49:22 +0000 Thin web server (v1.8.1 codename Infinite Smoothie)
2022-06-02 14:49:22 +0000 Maximum connections set to 1024
2022-06-02 14:49:22 +0000 Listening on 0.0.0.0:1234, CTRL+C to stop
[authn-k8s test server] Handling request: GET path=/some/path/api/v1
[authn-k8s test server] Handling request: GET path=/some/path/api/v1/namespaces/default/pods/bash-8449b79d7-c2fwd
[authn-k8s test server] Handling request: GET path=/some/path/api/v1/namespaces/default/pods/bash-8449b79d7-c2fwd/exec?container=bash&stderr=true&stdout=true&stdin=true&command=sh
[authn-k8s test server] Web socket event for request: [:open]
[authn-k8s test server] Web socket event for request: [:message, 0]
[authn-k8s test server] Web socket event for request: [:close, 1000, ""]
[authn-k8s test server] Handling request: GET path=/some/path/api/v1
[authn-k8s test server] Handling request: GET path=/some/path/api/v1/namespaces/default/pods/bash-8449b79d7-c2fwd
      client successfully authenticates

Finished in 2.33 seconds (files took 5.37 seconds to load)
1 example, 0 failures
```

### Desired Outcome

*Please describe the desired outcome for this PR.  Said another way, what was
the original request that resulted in these code changes?  Feel free to copy
this information from the connected issue.*

### Implemented Changes

*Describe how the desired outcome above has been achieved with this PR. In
particular, consider:*

- _What's changed? Why were these changes made?_
- _How should the reviewer approach this PR, especially if manual tests are required?_
- _Are there relevant screenshots you can add to the PR description?_

### Connected Issue/Story

Resolves #[relevant GitHub issue(s), e.g. 76]

CyberArk internal issue link: [insert issue ID]()

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

jvanderhoof commented:
@doodlesbykumbi, I really like the direction you're going with this. Two larger suggestions:

1. Allow user to define the request/response pattern so we can define various good/bad responses for any endpoint.

   Maybe something like this (pseudocode):

   ```ruby
   AuthnK8sTestServer.run(
     subpath: "/some/path",
     bearer_token: "bearer token",

     # request/response hash to define the supported requests/responses for this test
     responses: {
       'api/v1': 'good:api.v1.json',
       'api/v1/namespaces/default/pods/': 'bad:api.v1.getpod.json'
     }
   )
   ```

2. Wrap the web socket start/stop behavior so we don't need to worry about cleanup.

      Maybe something like this (pseudocode):
      ```ruby
      class AuthnK8sTestServer  
         ...
         # Allow `#run` to accept a block so we can handle startup/shutdown once the mock server is no longer needed.
         def self.run(...)
          # TODO: find out how to get random ports
          test_server = self.new(...)
          server = Rack::Handler::Thin.run(test_server, :Port => 1234, :Host => "0.0.0.0") do |server|
            Thread.current[:authn_k8s_test_server] = test_server
            AuthnK8sTestServer.log("Server running with port=#{server.port}, @subpath=#{server.app.subpath}")
          end
          yield(test_server)
          server.stop
         end
         ...
      end
      ```

      which would allow this to be called from the RSpec/Cucumber test like:
      ```ruby
      AuthnK8sTestServer.run(
        subpath: "/some/path", 
        bearer_token: "bearer token",
        # request/response hash to define the supported requests/responses for this test
        responses: {
          'api/v1': 'good:api.v1.json',
          'api/v1/namespaces/default/pods/': 'bad:api.v1.getpod.json'
        }
      ) do |server|
        signed_cert = server[:authn_k8s_test_server].copied_content
        # Authenticate request
        authn_k8s_authenticate(authenticator_id, account, host_id: test_app_host, signed_cert_pem: signed_cert.to_s)

        # Assertions
        expect(response).to be_ok
        token = Slosilo::JWT.parse_json(response.body)
        expect(token.claims['sub']).to eq("host/#{test_app_host}")
        expect(token.signature).to be
        expect(token.claims).to have_key('iat')
      end
      ```

doodlesbykumbi commented on Gemfile:
```diff
@@ -104,6 +105,7 @@ group :development, :test do
   gem 'spring-commands-cucumber'
   gem 'spring-commands-rspec'
   gem 'table_print'
+  gem 'thin'
```
Good point. Done

doodlesbykumbi commented:
@jvanderhoof Thanks for the feedback. Comments addressed

1. Removed `thin` in favour of `puma`
2. Changed the interface for running the test server to use the block so that we could incorporate some cleanup logic

Re:
> Allow user to define the request/response pattern so we can define various good/bad responses for any endpoint.

I wasn't able to address this now, as it requires some further design thinking. I'd like to skip it in this PR and revisit at a later date. I think the patterns that are currently hardcoded are enough to cover the test cases we're currently interested in.

john-odonnell commented on spec/support/authn_k8s/authn_k8s_test_server.rb:
```diff
@@ -0,0 +1,152 @@
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:api.v1.getnamespaces.json")] ]
+        elsif req.path.start_with?("#{subpath}/api/v1/namespaces")
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("bad:api.v1.getnamespaces.json")] ]
+        else
+            [ 200, {'Content-Type' => "application/json"}, ['{}'] ]
```
Might be a nitpick, but as we're enumerating all the endpoints the test case should need, maybe our fallback `else` condition should return a `40x` error instead of `200` response.

doodlesbykumbi replied:
This is a fair point. The issue I faced is that clients for Kubernetes hit a whole list of endpoints for (I think) service discovery purposes at initialisation time. Having the endpoints in the else branch return 200 overcame this issue. There's probably a better way.

I'll leave a comment on the else branch and then we can make improvements during the upcoming iterations on the test server.

john-odonnell reviewed (commented):
I'm enjoying this new testing strategy. Left some comments - nothing major or breaking.

doodlesbykumbi commented on app/domain/authentication/authn_k8s/execute_command_in_container.rb:
```diff
@@ -113,7 +113,7 @@ def verify_error_stream_is_empty
       def ws_exec_url
         api_uri = kube_client.api_endpoint.clone # contains /api path prefix
-        api_uri.scheme = "wss"
+        # api_uri.scheme = "wss"
```
With the current state of the PR, we stopped enforcing `wss`. Instead we accepted either `https` or `wss` as a signal to create a secure connection.

doodlesbykumbi replied:
With that in mind I'm happy to remove the commented out code

doodlesbykumbi commented on spec/controllers/authenticate_controller_authn_k8s_spec.rb:
```diff
@@ -0,0 +1,224 @@
+#   end
+# )
+
+def set_variable_value(resource_id, value)
+  post("/secrets/#{account}/variable/#{resource_id}", env: admin_request_env.merge({ 'RAW_POST_DATA' => value }))
+end
```
Good catch!

doodlesbykumbi commented on spec/controllers/authenticate_controller_authn_k8s_spec.rb:
```diff
@@ -0,0 +1,224 @@
+  set_variable_value("#{service_id}/kubernetes/service-account-token", service_account_token)
+end
+
+def fake_authn_k8s_login(account, service_id, host_id:)
```
It's fake in the sense that there's no call to the server. We simulate, in-line, the expected behaviour of the server.

john-odonnell reviewed (approved):
LGTM! Thanks for the background info
